tests/nilcheck/ref0.nim

```diff
@@ -0,0 +1,248 @@
+    it is nil by default on line 135:0
+  ref0.nim(139, 3) Warning: Cannot prove that 'result' is initialized. This will become a compile time error in the future. [ProveInit]
+  ref0.nim(138, 1) Error: return value is nil
+  ref0.nim(147, 1) Error: return value might be nil
+  ref0.nim(174, 9) Error: can't deref b.refField it might be nil
+    it has ref type on line 174:8
+  ref0.nim(183, 9) Error: can't deref b.refField it is nil
+    assigns a value which might be nil on line 181:19
+  ref0.nim(192, 9) Error: can't deref a[0] it might be nil
+    it has ref type on line 192:8
+  ref0.nim(198, 11) Error: can't deref a[b] it might be nil
+    isNil on line 196:13  a subexpression changed: we can't prove it isn't nil now on line 197:8
+  ref0.nim(216, 10) Error: can't deref a it might be nil
+    assigns a value which might be nil on line 214:6
+  ref0.nim(226, 10) Error: can't deref a it might be nil
+    assigns a value which might be nil on line 221:6
+  ref0.nim(237, 10) Error: can't deref b it might be nil
+    assigns a value which might be nil on line 234:6
+  ref0.nim(240, 8) Error: can't deref b it might be nil
+    assigns a value which might be nil on line 234:6
+  ref0.nim(244, 10) Error: can't deref a it might be nil
+    assigns a value which might be nil on line 243:6
+'''
+
+
+
+"""
+import tables
+
+type
+  A* = ref object
+    a*: float
+    b*: float
+
+  B* = ref object
+    refField*: A
+
+var a*: A
+
+proc e(a: A) = # A is Ref
+  echo a.a > 0 # can't deref a: it might be nil  
+
+proc e2(a: A) =
+  echo not a.isNil and a.a > 0
+
+proc e3(a: A) =
+  if a.isNil:
+    echo a.a # can't deref a: it is nil
+  else:
+    echo a.a
+
+proc e4(a: A, b: int) =
+  var a2 = a
+  if b == 0:
+    a2 = A()
+  echo a2.a # can't deref a2: it might be nil
+
+proc e5(a: A, b: int) =
+  if a.isNil:
+    echo 0
+  else:
+    echo a.a
+
+proc e6(a: A, b: int) =
+  var e = a
+  if e.isNil:
+    e = A()
+  echo e.a
+
+proc e7(a: A) =
+  var b = A()
+  for i in 0 .. 5:
+    echo b.a # can't deref b: it might be nil
+    if i == 2:
+      b = a
+  echo b.a
+
+proc f1: A =
+  nil
+
+proc f2 =
+  var a = f1()
+  echo a.a # can't defef a: it might be nil
+
+proc f3: A not nil =
+  A()
+
+proc f4 =
+  var a = f3()
+  echo a.a
+
+proc f5(a: var A) =
+  a = nil
+
+proc f6 =
+  var a = A()
+  echo a.a
+  f5(a)
+  echo a.a
+
+
+
+proc f7(a: A) =
+  case a.isNil:
+  of true: echo a.a
+  of false: echo 0
+
+proc f8(a: A): A =
+  echo result.a # can't deref result
+
+proc f9: A not nil =
+  echo result.a # can't deref result
+
+proc f10: A not nil = # return value is nil
+  discard 
+
+proc f11(a: int): A not nil = # why does it work?? ah nim adds result
+  if a == 0:
+    A()
+  else:
+    A()
+
+proc f12(a: int): A not nil = # return might be nil
+  if a == 0:
+    nil
+  else:
+    A()
+
+proc f13(a: int): A =
+  if a == 0:
+    nil
+  else:
+    A()
+
+proc f14(a: int): A not nil=
+  if a == 0:
+    return A()
+  else:
+    result = A()
+
+
```

Comment:
I assume you have these lines as "buffers" in order to be able to fix the line numbers in the error message above.

To solve this problem, I usually use the following process:

1) Write the code as properly formatted as possible.
2) Run the compiler to get the list of errors and it put in the spec above.
3) Re-run the compiler to produce the same list, but with corrected line numbers.
4) Copy-paste the new list of errors over the old one (the number of lines will stay the same and everything will be correct now)

Reply:
yeah, I started doing something like that in the later stages, it is a better approach(generally applicable to a lot of testing situations: run X which you know produces expected results and bootstrap tests from it)